MVVM/ViewModel/ZedMonitor.cs

```diff
@@ -0,0 +1,91 @@
+using Godot;
+using MQTTnet;
+using MQTTnet.Internal;
+using OpenCvSharp;
+using RoverControlApp.Core;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using System.Text.Json;
+using RoverControlApp.MVVM.ViewModel;
+
+namespace RoverControlApp.MVVM.ViewModel;
+
+public partial class ZedMonitor : Panel
+{
+	//Sprites for gyro visualisation
+    [Export]
+	Sprite2D pitchVisualisation;
+    [Export]
+    Sprite2D rollVisualisation;
+
+	//Labels for gyro display
+	[Export]
+	Label pitchDisplay;
+    [Export]
+	Label rollDisplay;
+    [Export]
+	Panel errorDisplay;
+
+    public event Func<MqttClasses.ZedImuData?, Task>? GyroscopeChanged;
+
+	MqttClasses.ZedImuData ?Gyroscope;
+
+    public Task OnGyroscopeChanged(string subTopic, MqttApplicationMessage? msg)
+	{
+		if (MainViewModel.Settings?.Settings?.Mqtt.TopicZedImuData is null || subTopic != MainViewModel.Settings?.Settings?.Mqtt.TopicZedImuData)
+			return Task.CompletedTask;
+        if (msg is null || msg.PayloadSegment.Count == 0)
+        {
+            MainViewModel.EventLogger?.LogMessage($"ZedMonitor Error: Empty payload");
+            return Task.CompletedTask;
+        }
+
+        try
+		{
+			Quaternion Quat = PullGyroscope(msg);
+			AngleUpdate(Quat);
+
+            return Task.CompletedTask;
+        }
+		catch (Exception e)
+		{
+			MainViewModel.EventLogger?.LogMessage($"ZedMonitor Error: {e.Message}");
+			return Task.CompletedTask;
+        }  
+    }
+    public Quaternion PullGyroscope(MqttApplicationMessage? msg)
+    {
+        try
+        {
+            Gyroscope = JsonSerializer.Deserialize<MqttClasses.ZedImuData>(msg.ConvertPayloadToString());
+            Quaternion Quat = new Quaternion((float)Gyroscope.orientation.x, (float)Gyroscope.orientation.y, (float)Gyroscope.orientation.z, (float)Gyroscope.orientation.w);
+            return Quat;
+        }
+        catch (Exception e)
+        {
+            GD.Print($"ZedMonitor Error (Something is wrong with json/deserialization): {e.Message}");
+            return new Quaternion();
+        }
+    }
+    public void AngleUpdate(Quaternion Quat)
+    {
+        double rollDeg = (180.0 / Math.PI) * Math.Atan2(2 * (Quat.W * Quat.X + Quat.Y * Quat.Z), 1 - 2 * (Quat.X * Quat.X + Quat.Y * Quat.Y));
+        double pitchDeg = (180.0 / Math.PI) * Math.Asin(2 * (Quat.W * Quat.Y - Quat.Z * Quat.X));
+        CallDeferred("DisplayUpdate", rollDeg, pitchDeg);
+    }
+    private void DisplayUpdate(double rollDeg, double pitchDeg)
+	{
+        if(rollDeg == 0 && pitchDeg == 0)

```

Comment:
Thought that both doubles being perfectly 0 is to improbable to take into account, but you are right. Fixed it in [950684f](https://github.com/wisniax/RoverControlApp/pull/10/commits/950684fee043a0b4565aeeb75a3627eee9c0262f). I also added a feature both showing time since last update to the user and showing information about data being outdated after a ceratain amount of time (setable in godot editor). [1072d25](https://github.com/wisniax/RoverControlApp/pull/10/commits/1072d25d49f58a91c64b48c5589f37cbd8cc9d30) Let me know if it's fine

Reply:
To be fair, it is improbable :)
### It's almost fine. ###
If `errorTime` can't be divided by `timerInterval` without remainder, error about gyro data being too old won't be logged (saved to .log file). 
> [MVVM/View/ZedMonitor.tscn](https://github.com/Wojtuz/RoverControlApp/blob/1431fceced6df88da1ca00c941195e6141e09b8c/MVVM/ViewModel/ZedMonitor.cs#L78-L92)

```cs
public void _on_timer_timeout()
{
    timeSLU += timerInterval;

    if (timeSLU == errorTime)
    {
        MainViewModel.EventLogger?.LogMessage($"ZedMonitor Error: gyro data >{errorTime} seconds old.");

    }
    if (timeSLU >= errorTime)
    {
        errorDisplay.Visible = true;
    }
    TimeUpdate(timeSLU);
}
```